### ISSUE TITLE
Move index.php and assets to /public folder

### ISSUE THREAD

**Issue by karser** [microweber/microweber#532]
Exposing all unrelated stuff to the web is the insecure solution. 
For example:
- http://demo.microweber.com/vendor/
- http://demo.microweber.com/composer.lock

Hackers know the exact versions of installed dependencies, they could analyze them and find a vulnerability, for example, how to upload a file or exec the code.

The web root should be a separate folder that contains index.php, assets, etc. Btw, it's done in laravel by default https://github.com/laravel/laravel/tree/master/public

**Comment by peter-mw** [comment]
Thanks for the pull request

**Comment by ekordev** [comment]
can't you just restrict access to these files?

**Comment by marcnewton** [comment]
I was just checking out this project and that is the first thing I noticed is how the public dir was flattened into the root creating one gigantic security vulnerability.

Good job there Microweber on taking Laravel and making it totally insecure.

I assume somebody didn't understand how you where supposed to connect storage to public directory so they butchered it.
Simply as part of your deployment scripts you create a symlink in public by calling `php artisan storage:link` after `composer install --no-dev` has been called.

The public folder must be put back in it's rightful place!

**Comment by peter-mw** [comment]
Hi, @marcnewton 

Thanks for pointing out this issue, we could use some improvement if we find some good way.

At the moment the folders are secured with .htaccess, same way as other cms do. 

The general problem is the mechanics of shared hosting deployment. 

The user just uploads stuff in `public_html` also many automated installation scripts like Softaculous are working the same way. 

Maybe for the advanced user we can offer a way to remove non public folders. 

Do you have any ideas how this can be done ? 

We will check on `php artisan storage:link`

**Comment by marcnewton** [comment]
@peter-mw, Their lays a huge problem with such hosting solutions.
One major issue is that many hosts actually have .htaccess overrides disabled by default so the file is there but it does nothing, a .htaccess file is considered to be a vulnerability by today's standards.
Secondly when a novice wants to add redirect rules to a .htaccess, often happens, many of the posted examples out there that people will copy, actually cause the rule set to end prematurely so the protect directories rules don't get applied.

It is far too easy for novices to deploy a eCommerce platform that puts consumers and their data at risk.

I have been asked if I could review this platform to help build a secured deployment with nginx so I will take a look to see if I can devise some solution for apache and nginx and those dirty dir hosts.

**Comment by peter-mw** [comment]
Hi, @marcnewton 

We will really appreciate some help on this issue.
 
So far the only folder that needs public access is the `userfiles` folder. The best solution is if we can have only `index.php`, `.htaccess` and `userfiles` in the `public_html` dir. 

Simple solution for nginx for is to exclude access all other folders. 

The good solution is to make a parameter in the [install command](https://github.com/microweber/microweber/blob/dev/src/Microweber/Commands/InstallCommand.php) to move all folders out of the public dir and be able to do something like `php artisan microweber:install admin@site.com admin password storage/database1.sqlite microweber microweber nopass sqlite -p site_ -t dream -secure`

**Comment by karser** [comment]
Hi folks!

>The general problem is the mechanics of shared hosting deployment.

Maybe it could be fixed at least for docker, during the docker image build? e.g:
```
WORKDIR /var/www/microweber
COPY . ./
COPY ./index.php ./favicon.ico ./public/
RUN rm ./index.php ./favicon.ico
```
where `/var/www/microweber/public` is the document root
wdyt?

**Comment by marcnewton** [comment]
@peter-mw Best theory I have so far is to move the entirety of Laravel up one directory into a DOT prefix DIR as many of the shared hosts are configured by default to protect DOT notation directories and files from public access, at the very least it would mean only having to ensure the end user protects one single directly instead of dozens.

I would then either:
A) Have an `artisan` script `Symlink` the files in the original Laravel `/.laravel/public` dir and the `/.laravel/storage` directory.

or

B) `Symlink` the `storage` dir and have a copy of the /public content in /public_html but import/require `/.laravel/public/index.php`:

```
/public_html/.laravel/{app,vendor,database,resources,public,etc...}
/public_html/index.php > require/import('./.laravel/public/index.php');
/public_html/.htaccess 
/public_html/favicon.ico
```

For DOCKER deployments, could simply deploy everything as is out of the `/.laravel/` to `root` giving them the vanilla directory layout with `/public` and run the native symlink command `php artisan storage:link` .

In regards to `/public_html/index.php`  that should reference the original  `/public/index.php` I can't remember if a `require()` or an `import()` would serve best for ensuring the rest of cascading import function correctly for the PSR `/vendor/autoloader`, one of the two should give the desired result.

This should serve both technical and none technical deployments.

I have yet to test this theory my self but I have done similar things a very long time ago, should work...

**Comment by marcnewton** [comment]
@peter-mw I spent 6 hours going over the source code, In all honesty in it's current state I would just start again porting over to a new Microweber2 project.

The architecture in it's current form does not boast well for continuous integration with the Laravel update cycle. There are a lot of strange hacks that decouple this from being a pure Laravel based project.

Too much reliance on 3rd party packages that have crisscrossed version requirements making for a huge upgrade lead time, There are some unnecessary package usages.

Views and theme concept could have been done better and not deviate from the use of blade templates, an opportunity to make better use view name spaces was clearly missed..
I see the late a very late registration being done in `DefaultController`.

```php
// ThemeServiceProvider or BeforeMiddleware could have registered desired view configurations.

View::addNamespace('backend', base_path() . '/resources/admin/views');
View::addNamespace('frontend', base_path() . '/resources/themes/' . config('microweber.frontend.theme') . '/views');

...

// calling in a controller
view('frontend::checkout');
```

Ambiguous coding practices or lazy helpers, a heavy use of process offloads, while they save dev time that can lead to various security risks especially when combined with mass-assignable fields which are used often in this source..

**Comment by peter-mw** [comment]
Hi @marcnewton ,

We plan to do refactoring for version 2 which will look a lot more like actual laravel app. 

The code at its current state has been around before we migrated to Laravel and thats why its looks like a layer glued to it.  

There is a way to write code in with Blade and Models,  http://docs.microweber.com/guides/modules_laravel.md  , just its lot of work to rewrite all existing code.  

For the next resales will upgrade to Laravel 6 and try to solve the public folder problem and deal with some existing bugs. 

For real refactoring to be done we need to split the code on packages and do refactoring from there, but this will be after upgrading to Laravel 6

**Comment by barbosa89** [comment]
A simple option could be:

1. Delete the public_html folder from the shared hosting.
2. Create a symbolic link named public_html to point to Laravel public folder (public_html --> laravel-app/public), so the public_html folder is replaced for a symbolic link. 

You could run a script during the installation

**Comment by ekordev** [comment]
> A simple option could be:
> 
> 1. Delete the public_html folder from the shared hosting.
> 2. Create a symbolic link named public_html to point to Laravel public folder (public_html --> laravel-app/public), so the public_html folder is replaced for a symbolic link.
> 
> You could run a script during the installation

If they're using shared hosting they most likely will not be able to delete the public_html file plus most shared host don't allow scripts that create symbolic links.

**Comment by barbosa89** [comment]
Yes, SSH access is required, I have performed this method in A2 Hosting and Hostinger with excellent results, I have implemented Laravel applications without problems.

**Comment by ekordev** [comment]
If you have hostinger try running `php artisan storage:link`  through ssh. It won't let you, a script attempting to create a symbolic link for public_html in public would get the same response. I currently have 6 laravel websites on my Hostinger shared hosting account so I know it's possible using work arounds but still has limitations

**Comment by ekordev** [comment]
To clarify yes you can create symbolic links using ln -s through ssh for hostinger but I've tried deleting public_html and creating a link to it in public but it didn't work I received a server not found response

**Comment by barbosa89** [comment]
You need to do two things:

1. Delete public_html folder in the user folder:

Route: /home/username/public_html

After, you need create a symlink to project:

ln -s /home/username/path_to/laravelapp/public /home/username/public_html

2. Delete public_html folder in the domain name folder:

Route: /home/username/domains/domain_name/public_html

After, you need create a symlink to project:

ln -s /home/username/path_to/laravelapp/public /home/username/domains/domain_name/public_html

Enjoy.

I tried to run the `php artisan storage:link` command and it works fine in Hostinger
